tqftpserv: ACK during write based on wsize

bhaskarv79 opened the pull request:
Client expects ACK in WRQ request based on wsize.  Fix is added to wait for wsize packets and then ACK, wsize buffers are then written to the file.

lumag commented on tqftpserv.c:
```diff
@@ -138,7 +143,7 @@ static int tftp_send_oack(int sock, size_t *blocksize, size_t *tsize,
 	}
 
-	if (tsize && *tsize != -1) {
+	if (tsize && !*tsize) {
```
This chunk and the next one look like an independent fix, please split them to a separate commit.

quic-bhaskarv replied:
this is a bug, atoi can return 0 for incorrect value.  So I assigned -1 instead of 0.
I can move to next patch.

quic-bhaskarv replied:
Noted and Updated.

lumag commented on tqftpserv.c:
```diff
@@ -46,6 +47,10 @@ struct tftp_client {
 	unsigned int timeoutms;
 	off_t seek;
+	size_t cur_wsize;
+	uint8_t *rw_buf;
```
I don't see where it is being assigned.

lumag commented on tqftpserv.c:
```diff
@@ -522,24 +530,35 @@ static int handle_writer(struct tftp_client *client)
 
 	payload = len - 4;
+	buf += 4;
+	memcpy(client->rw_buf+client->cur_buf_loc, buf, payload);
```
Missing a check that rw_buf has enough space for memcpy

lumag commented on tqftpserv.c:
```diff
@@ -522,24 +530,35 @@ static int handle_writer(struct tftp_client *client)
-		printf("[TQFTP] failed to write data\n");
-		return -1;
+
+	if (client->rw_buf_size <= 0 || payload < client->blksize) {
```
If it is less than 0, it means that you've overritten something after the buffer.

quic-bhaskarv replied:
It shouldn't be <0, I will remove < check.

lumag reviewed (changes requested):
- Rewrite the commit message in the imerative style
- Drop Gerrit tag
- Add SoB

lumag commented on tqftpserv.c:
The size doesn't decrease when you add more data. Please rename `rw_buf_size` accordingly.

lumag commented on tqftpserv.c:
Okay, we've come to the main part. Is there a reason to buffer all the data to write it in one chunk? Wouldn't it be better to write it as it was written before and just change when the ACK is being sent (as you've stated in the commit message)?

quic-bhaskarv replied:
Yes, two reasons:
1. clients expect ACK after wsize packets.  If it receives for every packet then it will assume server is not following wsize option and can decide to restart the connection.
2. It is possible that the client/server may not have received packets and client can decide to retransmit.  In that case the buffer written to file can be corrupted.

In this patch I have introduced sending ACK after receiving wsize packets, this will handle 1st case.  I will submit another patch to handle 2nd case, where server will track the block# and ensure that wsize packets are received in order.

lumag replied:
The point 1 doesn't really answer my question (you can still be sending ACKs once for each wsize rather than each packet).  Point 2 doesn't seem relevant and (as you wrote) it is not covered by the current patch.

quic-bhaskarv replied:
I can add the code for the second part, I thought it should be created as a separate patch as the logic will be new.  I will push another commit that introduces this logic.

lumag reviewed (changes requested):
Other than the review comments, please rework your commits. Separate your changes into atomic commits.

quic-bhaskarv commented on tqftpserv.c:
```diff
@@ -322,6 +323,18 @@ static void handle_rrq(const char *buf, size_t len, struct sockaddr_qrtr *sq)
 	client->timeoutms = timeoutms;
 	client->seek = seek;
 
+	client->blk_buf = calloc(1, blksize+4);
```
I will create new patch to introduce blk_buf and use it in handle_{reader,writer}

quic-bhaskarv commented on tqftpserv.c:
```diff
@@ -419,15 +452,15 @@ static int handle_reader(struct tftp_client *client)
 	uint16_t block;
 	uint16_t last;
-	char buf[128];
+	uint8_t *buf = client->blk_buf;
```
will raise new change

quic-bhaskarv commented on tqftpserv.c:
```diff
@@ -510,7 +543,7 @@ static int handle_writer(struct tftp_client *client)
 	/* Drop unsolicited messages */
 	if (sq.sq_node != client->sq.sq_node ||
-	    sq.sq_port != client->sq.sq_port)
+			sq.sq_port != client->sq.sq_port)
```
reverted

quic-bhaskarv commented:
Please let me know if there are feedback comments for these patches.

lumag commented:
> Please let me know if there are feedback comments for these patches.

There are 5 commits in the PR, 4 of which share the same commit message. Please squash them and restrucutre in a logical way. Each commit should cover exactly one logical set of changes.

quic-bhaskarv commented:
I have squashed the gerrits to two.  Please review.

quic-bhaskarv commented:
Please let me know if there any more comments.

lumag commented on tqftpserv.c:
```diff
@@ -522,15 +541,46 @@ static int handle_writer(struct tftp_client *client)
+	buf += 4;
+
+	/* Check if we recieved expected block */
+	if (block != client->blk_expected) {
```
This is not related to wsize ACK handling. Please make all the commits perform one logical change.

lumag replied:
Ensuring the sequence is a separate change, even if we don't send ACKs once per wsize.

quic-bhaskarv replied:
I created multiple patches and separated based on the logic.

lumag commented on tqftpserv.c:
```diff
@@ -46,6 +46,11 @@ struct tftp_client {
 	unsigned int timeoutms;
 	off_t seek;
+
+	uint8_t *rw_buf;
```
I still don't see a need for a wsize buffer to handle the ACK field. Please change your commit message to describe the reasons for the change and for architectural decisions (in other words, describe 'why', not 'what').

lumag replied:
I'd probably code it differntly: write pieces to the file, but store the last "good" offset. If we send NACK, then rollback to that position and restart.

quic-bhaskarv replied:
we can do that way as well, however this server is used on NAND storage media as well.  Seeking and writing may be heavy operation on unmanaged storage media.

Please review latest code and let me know.

lumag commented on tqftpserv.c:
```diff
@@ -47,6 +47,7 @@ struct tftp_client {
 	unsigned int timeoutms;
 	off_t seek;
 
+	uint8_t *blk_buf;
```
Please start your commit message by describing the problem. What is currently broken and why.

lumag replied:
It is better to have a big commit message than the one not describing the issue.

bhaskarv79 committed:
tqftpserv: Set default wsize to 1 if not specified in WRQ/RRQ

Ensure the default wsize value is set to 1 when it is not provided
in a Write Request (WRQ) or Read Request (RRQ).

Signed-off-by: Bhaskar Valaboju <quic_bhaskarv@quicinc.com>

bhaskarv79 committed:
tqftpserv: Allocate memory in client instance to hold blksize buffer

Allocate memory in the client instance to support transmission and
reception of TFTP data packets sized according to the negotiated
blksize option. This buffer ensures proper handling of variable-length
payloads during transfer operations.

Signed-off-by: Bhaskar Valaboju <quic_bhaskarv@quicinc.com>

bhaskarv79 committed:
tqftpserv: Allocate wsize buffer in TFTP client instance

Allocate a buffer within the client instance that can hold
the wsize value when reading from or writing to the TFTP client.

Signed-off-by: Bhaskar Valaboju <quic_bhaskarv@quicinc.com>

bhaskarv79 committed:
tqftpserv: Use blksize buffer in send/recvfrom for TFTP packets

Use the blksize-sized buffer from the client instance for send
and recvfrom operations during TFTP packet handling.

This eliminates repeated memory allocation in handler_reader
and supports receiving full blksize payloads in handler_writer.

Signed-off-by: Bhaskar Valaboju <quic_bhaskarv@quicinc.com>

bhaskarv79 committed:
tqftpserv: Cache wsize buffers in client instance

The client sends data packets of blksize size and waits for an ACK
after sending wsize packets. If packets are not received in sequence,
the server should send NACK, prompting the client to resend the packets
starting from beginning of the window.

For example, with wsize = 10 and blksize = 5000, the client sends
blocks [201–210] and waits. If all are received, the server replies
with ACK; otherwise, it sends NACK, and the client restarts from block 201.

Use rw_buf to cache received blocks and track the next expected block
using blk_expected. Reset the expected block to the start of the window
if out-of-sequence packets are detected.

Once all buffers are received, commit the cached buffers to the file.

Signed-off-by: Bhaskar Valaboju <quic_bhaskarv@quicinc.com>
